Tools/Cli-LoRa-Device-Provisioning/Program.cs

```diff
@@ -370,6 +372,89 @@ private static async Task<bool> UploadCertificateBundleAsync(string certificateB
             Task CleanupAsync() => blobClient.DeleteIfExistsAsync();
         }
 
+        private static async Task<bool> RunUpgradeFirmwareAndReturnExitCodeAsync(UpgradeFirmwareOptions opts)
+        {
+            if (!configurationHelper.ReadConfig())
+                return false;
+
+            if (!File.Exists(opts.FirmwareLocation))
+            {
+                StatusConsole.WriteLogLine(MessageType.Error, "Firmware upgrade file does not exist at the specified location.");
+                return false;
+            }
+
+            if (!File.Exists(opts.DigestLocation))
+            {
+                StatusConsole.WriteLogLine(MessageType.Error, "Digest of the firmware upgrade does not exist at the specified location.");
+                return false;
+            }
+
+            if (!File.Exists(opts.ChecksumLocation))
+            {
+                StatusConsole.WriteLogLine(MessageType.Error, "CRC32 checksum of the signature key does not exist at the specified location.");
+                return false;
+            }
+
+            // Upload firmware file to storage account
+            var success = await UploadFirmwareAsync(opts.FirmwareLocation, opts.StationEui, opts.Package, async (firmwareBlobUri) =>
+            {
+                var twin = await IoTDeviceHelper.QueryDeviceTwin(opts.StationEui, configurationHelper);
+
+                if (twin is null)
+                {
+                    StatusConsole.WriteLogLine(MessageType.Error, "Device was not found in IoT Hub. Please create it first.");
+                    return false;
+                }
+
+                // Update station device twin
+                twin.Properties.Desired[TwinProperty.Cups][TwinProperty.FirmwareVersion] = opts.Package;
+                twin.Properties.Desired[TwinProperty.Cups][TwinProperty.FirmwareUrl] = firmwareBlobUri;
+                twin.Properties.Desired[TwinProperty.Cups][TwinProperty.FirmwareKeyChecksum] = File.ReadAllText(opts.ChecksumLocation, Encoding.UTF8);
```

Comment:
This should not be saved as string, please try parse it as uint and save it as such